sklearn/base.py

```diff
@@ -555,6 +556,33 @@ def fit_transform(self, X, y=None, **fit_params):
             # fit method of arity 2 (supervised transformation)
             return self.fit(X, y, **fit_params).transform(X)
 
+    @property
+    def n_features_out_(self):
+        return self._n_features_out
+
+    @n_features_out_.setter
+    def n_features_out_(self, val):
+        self._n_features_out = val
+
+
+class ComponentsMixin:
+    @property
+    def n_features_out_(self):
+        if hasattr(self, 'n_components_'):
+            # n_components could be auto or None
+            # this is more likely to be an int
```

Comment:
We can also include the `isinstance(..., numbers.Integral)` check here to be sure.